File: 2025/Day 1/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// file, err := os.ReadFile("example.txt")
	file, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	fileContent := strings.TrimSpace(string(file))
	fileArray := strings.Split(fileContent, "\n")

	result1 := partOne(fileArray)
	// result2 := partTwo(fileArray)
	fmt.Printf("Part 1: %d\n", result1)
	// fmt.Println(result1 + result2)
}

func partOne(fileArray []string) int {
	position := 50
	zeroCount := 0

	for i := 0; i < len(fileArray); i++ {
		instruction := fileArray[i]
		direction := instruction[0]
		initialDistance, _ := strconv.Atoi(instruction[1:])

		distance := initialDistance % 100

		if direction == 'L' {
			position -= distance
			if position < 0 {
				position = 100 + position
			}
		} else {
			position += distance
			if position > 99 {
				position = position - 100
			}
		}

		if position == 0 {
			zeroCount++
		}
	}

	return zeroCount
}

func partTwo(fileArray []string) int {
	position := 50
	zeroCount := 0

	for i := 0; i < len(fileArray); i++ {
		instruction := fileArray[i]
		direction := instruction[0]
		distance, _ := strconv.Atoi(instruction[1:])

		for distance > 0 {
			if direction == 'L' {
				position--
			} else {
				position++
			}
			if position == -1 {
				position = 99
			} else if position == 100 {
				position = 0
			}
			if position == 0 {
				zeroCount++
			}
			distance--
		}
	}

	return zeroCount
}

File: 2025/Day 1/main.ts
import { readFileSync } from 'fs';

// const file = readFileSync('example.txt', 'utf8').trim();
const file = readFileSync('input.txt', 'utf8').trim();

const fileArray = file.split('\n');

function partOne() {
    let position = 50
    let zeroCount = 0


    for (let i = 0; i < fileArray.length; i++) {
        const instruction = fileArray[i];
        const direction = instruction[0]
        const initialDistance = parseInt(instruction.slice(1))

        const distance = initialDistance % 100;

        if (direction === 'L'){
            position -= distance;
            if (position < 0) {
                position = 100 + position;
            }

        } else {
            position += distance;
            if (position > 99) {
                position = position - 100;
            }
        }

        if (position === 0) {
            zeroCount++;
        }
    }

    return zeroCount;
}

function partTwo() {
    let position = 50
    let zeroCount = 0

    for (let i = 0; i < fileArray.length; i++) {
        const instruction = fileArray[i];
        const direction = instruction[0]
        let distance = parseInt(instruction.slice(1))

        while(distance > 0) {
            if (direction === 'L') {
                position--;
            } else {
                position++;
            }
            if(position === -1) {
                position = 99;
            } else if(position === 100) {
                position = 0;
            }
            if (position === 0) {
                zeroCount++;
            }
            distance--;
        }
    }

    return zeroCount;
}

const start1 = performance.now();
const result1 = partOne();
const elapsed1 = performance.now() - start1;
console.log(`Part 1: zero count is ${result1} (took ${elapsed1.toFixed(3)}ms)`);

const start2 = performance.now();
const result2 = partTwo();
const elapsed2 = performance.now() - start2;
console.log(`Part 2: zero passes is ${result2} (took ${elapsed2.toFixed(3)}ms)`);

console.log(result1 + result2);
